### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by andrewrynhard**
Signed-off-by: Andrew Rynhard <andrew@andrewrynhard.com>

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "bldr",
+	Short: "A brief description of your application",
```
Comment (andrewrynhard) [inline cmd/root.go]: Needs a description.

File: hack/test/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,2 @@
+FROM --platform=$TARGETPLATFORM alpine:3.9 as test
+COPY --from=docker.io/autonomy/golang:059383d-dirty /toolchain /toolchain
```
Comment (andrewrynhard) [inline hack/test/Dockerfile]: Make this generic with an `ARG`.

File: cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// Copyright © 2019 NAME HERE <EMAIL ADDRESS>
```
Comment (bradbeam) [inline cmd/root.go]: nit, update copyright/license

File: internal/pkg/config/config.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	"gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	Repository string `yaml:"repository,omitempty"`
+	Username   string `yaml:"username,omitempty"`
+}
+
+func Open() (*Config, error) {
+	c := &Config{}
```
Comment (bradbeam) [inline internal/pkg/config/config.go]: not sure I care about it for initial PR, but we should add support for specifying the bldr.yaml location

File: internal/pkg/types/v1alpha1/build.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+		}
+		if err := step.test(p.Shell); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	return nil
+}
+
+func environment() error {
```
Comment (bradbeam) [inline internal/pkg/types/v1alpha1/build.go]: this func is probably fine for now, but we should look at either setting these as defaults that can be overridden or moved to bldr.yaml

File: internal/pkg/types/v1alpha1/build.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+		case "armv7l":
+			os.Setenv("BUILD", "armv7-linux-musl")
+			os.Setenv("HOST", "armv7-linux-musl")
+		default:
+			result = multierror.Append(result, xerrors.Errorf("unsupported build platform: %q", arch))
+			return
+		}
+	}
+
+	basic()
```
Comment (bradbeam) [inline internal/pkg/types/v1alpha1/build.go]: this pattern with inline funcs seems odd/maybe even not right. I think it'd be clearer to just have these without the inline func as a series of switch statements.

File: internal/pkg/types/v1alpha1/pack.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		"--tag=" + strings.Join([]string{p.Config.Repository, p.Config.Username, p.Name}, "/") + ":" + p.Metadata.Container.Image.Tag,
+		"--file=" + tmpfile.Name(),
+		".",
+	}
+
+	return cmd("docker", args...)
+}
+
+func (i Install) AsDockerRun() string {
+	return fmt.Sprintf("RUN apk add --no-cache %s", strings.Join(i, " "))
```
Comment (bradbeam) [inline internal/pkg/types/v1alpha1/pack.go]: fine for now but probably want to add a switch statement based on base image ( idk if regex or strings.contains )
```
switch 
case "alpine"
case "ubuntu"
case "centos"
...
```

File: internal/pkg/types/v1alpha1/build.go (hunk 6 above)
Comment (andrewrynhard) [reply internal/pkg/types/v1alpha1/build.go]: The inline funcs were just a way to help the readability. They really don't need to stand alone as individual funcs. The names of the func vars are hopefully descriptive and make it clear what they are for. Happy to implement another way, but looking for a way to make it clear that some are general, some are for build, and some are for target https://wiki.osdev.org/Target_Triplet.

File: internal/pkg/types/v1alpha1/pack.go (hunk 7 above)
Comment (andrewrynhard) [reply internal/pkg/types/v1alpha1/pack.go]: Absolutely. Cool if we add this on later? I agree 100%, just have my concerns with time.

**Comment by andrewrynhard** [comment]
I think this is in a state that should work for us.

**Pull request merged**